### PR TITLE
Implement URL sanitization

### PR DESCRIPTION
**Pull request by mprobst**

### CODE REVIEW

File: modules/@angular/compiler/src/template_parser.ts [hunk 1]
```diff
@@ -718,27 +720,35 @@ class TemplateParseVisitor implements HtmlAstVisitor {
     } else {
       if (parts[0] == ATTRIBUTE_PREFIX) {
         boundPropertyName = parts[1];
+        // NB: For security purposes, use the mapped property name, not the attribute name.
+        securityContext = this._schemaRegistry.securityContext(
```
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser.ts]: Might be better to only fill the `SecurityContext.STYLE` here, and calculate the other kinds in `_createElementPropertyAst` as we do `getMappedPropName` there already (i.e. there we calculate which property should be used for real).

**Comment by tbosch** [comment]
Looks good so far (didn't look at nitpicks...)

File: modules/@angular/compiler/src/template_parser.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/compiler/src/template_parser.ts]: Ah, never mind. This is right!

File: modules/@angular/compiler/src/schema/dom_element_schema_registry.ts [hunk 2]
```diff
@@ -254,6 +255,12 @@ export class DomElementSchemaRegistry implements ElementSchemaRegistry {
     }
   }
 
+  securityContext(tagName: string, propName: string): SecurityContext {
+    if (tagName === 'a' && propName === 'href') return SecurityContext.URL;
```
Comment (tbosch) [inline modules/@angular/compiler/src/schema/dom_element_schema_registry.ts]: I think we need to refactor this:
`SecurityContext.HTML` and `SecurityContext.URL` only make sense in a browser. I.e. you calculate them in the `DomElementSchemaRegistry`.

File: modules/@angular/compiler/src/view_compiler/property_binder.ts [hunk 3]
```diff
@@ -129,6 +130,33 @@ function bindAndWriteToRenderer(boundProps: BoundElementPropertyAst[], context:
   });
 }
 
+function sanitizedValue(boundProp: BoundElementPropertyAst, renderValue: o.Expression): o.Expression {
```
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/property_binder.ts]: Escaping should happen in the `DomRenderer`.

File: modules/@angular/core/src/security/sanitization_service.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
+import {Injectable} from '../di/decorators';
```
Comment (tbosch) [inline modules/@angular/core/src/security/sanitization_service.ts]: This sanitization service should live under `src/platform/dom` as it is DOM specific.

**Comment by tbosch** [comment]
Ok, how about the following change:
1. have a general abstract type `SecurityContext` with no properties
2. calculate the `SecurityContext` in the `DomSchemaRegistry`
3. pass the `SecurityContext` to `DomRenderer.setElementProperty(el, propName, propValue)`.

As context values, I think you don't need SecurityContext.STYLE as there is a specific method on the `DomRenderer.setStyle` to deal with styles.

I.e. the `SanitizationService` should also go into `src/platform/dom`.

**Comment by tbosch** [comment]
Ok, so I would propose this to separate the platform specific aspects:

Interfaces:

```
abstract class SchemaRegistry {
  securityContext(tagName: string, propName: string): CompileIdentifierMetadata
}

class BoundElementPropertyAst {
...
  securityContext: CompileIdentifierMetadata;
...
}

abstract class SanitizationService {
  sanitize(securityContext: any, value: any): any;
}

property_binder.ts:
function sanitizedValue(boundProp: BoundElementPropertyAst, renderValue: o.Expression): o.Expression {
  return ViewProperties.viewUtils.prop('sanitizer').callMethod('sanitize', 
      [o.importExpr(boundProp.securityContext), renderValue]);
}

```

Implementations:

```
class DomSchemaRegistry {
  securityContext(tagName: string, propName: string): CompileIdentifierMetadata {
    ...
    return new CompileIdentifierMetadata({
      name: 'DomSecurityContext.HTML',
      moduleUrl: 'angular2/src/platform/dom/DomSecurityContext',
      runtime: DomSecyrityContext.HTML
    });
  }
}

enum DomSecyrityContext {
  HTML,
  ...
}

class DomSanitizationService implements SanitizationService {
  santitize(securityContext: DomSecyrityContext, value: any) {
    switch (securityContext) {
      case DomSecurityContext.HTML: ...
    }
  }
}
```

**Comment by mprobst** [comment]
@tbosch:

I have concerns :-)
- `CompileIdentifierMetadata` can only be an import, not even an enum value. 
- It gets harder in the code generation phase to decide to not generate code at all - which seems highly attractive. Otherwise you end up with lots of superfluous calls to `sanitizer.sanitize(null, ...)`. Special casing a specific `null` value for `CompileIdentifierMetadata` and then special casing the property binder again for it feels dirty.
- `CompileIdentifierMetadata` is an artefact of the code generation phase. Pushing that into the schema manager seems like a stark layering violation.
- Having all renderers deal with a very specific implementation detail in the form of `CompileIdentifierMetadata` or similar over just a SecurityContext.NONE seems to create unnecessary complexity.
- This degenerates the APIs and implementation specifics on the template parsing/code generation layer into a heap of `any`s, making it that much harder to read and write.

But more importantly, this abstraction turns security concerns into a dynamic property, as opposed to something we can analyze statically. E.g. we want to fail statically when somebody writes `<a onclick="...">` and tell them to write `<a (click)=...>`, we want to generate specific code for the appropriate security invocations, to generate code that's fast and sensible and obviously correct. Or we can create another actor that gets queried by the template compiler, but then we have another abstraction that doesn't make sense for Native Script.

In general, if we wanted to create a template parsing/code generation layer that's completely platform agnostic but can still do these attractive goals, we'd have to have an abstraction for the template parsing, AST, and code generation with a specialization for Native Script vs DOM that wires through the security, so that Native Script can be oblivious to the concept existing. That's however substantial effort and design for which we don't have a good second use case right now, so we'll likely get it wrong if we try.

In any case, I'm unconvinced that that would be even worth it. Yes, it's a bit ugly that the word `SecurityContext` appears for Native Script, and that something in the template compiler talks about the DOM. But on the other hand, this just means Native Script has to return `SecurityContext.NONE` at one point, and bind a no-op implementation for `SanitizationService`.

That seems like a perfectly fine tradeoff in exchange for the ability to do static security analysis. The layering violation really is not that bad (IMHO much less than the `CompileIdentifierMetadata` thing!).

So I think we should continue with this general approach of having the template parsing/compiler layer be aware of the specific DOM SecurityContext values. If we decide to completely abstract code generation for the different platforms, we can still refactor it then.

**Comment by tbosch** [comment]
> CompileIdentifierMetadata can only be an import, not even an enum value.

It can, see e.g. [ViewType.HOST](https://github.com/angular/angular/blob/master/modules/angular2/src/compiler/view_compiler/constants.ts#L26)

> It gets harder in the code generation phase to decide to not generate code at all - which seems highly attractive. Otherwise you end up with lots of superfluous calls to sanitizer.sanitize(null, ...). Special casing a specific null value for CompileIdentifierMetadata and then special casing the property binder again for it feels dirty.

Yes, we should not call `santitizer.sanitize(null, ...)`. However I think that using a `null` result from the `SchemaRegistry` to indicate that we should not generate this code is not bad

> CompileIdentifierMetadata is an artefact of the code generation phase. Pushing that into the schema manager seems like a stark layering violation.
- `CompileIdentifierMetadata` is the base class for how we represent the type of directives, pipes, providers in the parsed `TemplateAst` already. I.e. our template parser (which builds up the semantic `TemplateAst` based upon the syntactic `HtmlAst`) already deals with them.
- `CompileIdentifierMetadata` is a nice way of referencing directives (and other values) in a way that works for interpretation mode (where we don't eval) as well as for codegen as it combines a runtime value with a representation of that value that is usable for codegen (module name + exported name, with a special case for enums allowing a `.` in the name).

> Having all renderers deal with a very specific implementation detail in the form of CompileIdentifierMetadata or similar over just a SecurityContext.NONE seems to create unnecessary complexity.
- The proposed change above does not require to change the `Renderer` API or any `Renderer` implementation (although my previous comment that I removed suggested this).

> This degenerates the APIs and implementation specifics on the template parsing/code generation layer into a heap of anys, making it that much harder to read and write.

I don't see a heap of `any`, except for the runtime interface of the `SanitizationService`. The value needs to be `any`, the proposal above only makes the `securityContext` also an `any` (compared to the implementation in this PR).

> But more importantly, this abstraction turns security concerns into a dynamic property, as opposed to something we can analyze statically. E.g. we want to fail statically when somebody writes <a onclick="..."> and tell them to write <a (click)=...>, we want to generate specific code for the appropriate security invocations, to generate code that's fast and sensible and obviously correct. 

The proposal above is only adding interfaces above the implementation in this PR, and instead of generating code that specifically calls `sanitizeHtml`, `sanitizeUrl`, ... it would call `sanitize(DomSecurityContext.HTM, ...)`, `sanitize(DomSecurityContext.URL, ...)`. This is still very readable (obviously correct) and regarding performance only one `switch` over the `SecurityContext` slower.

> Or we can create another actor that gets queried by the template compiler, but then we have another abstraction that doesn't make sense for Native Script.

We already have this "other actor" that gets queries to check which properties / elements exist in form of the `SchemaRegistry`, with the special case of the `DomSchemaRegistry`. This would be enough to report an error for `onclick`.

> In general, if we wanted to create a template parsing/code generation layer that's completely platform agnostic but can still do these attractive goals, we'd have to have an abstraction for the template parsing, AST, and code generation with a specialization for Native Script vs DOM that wires through the security, so that Native Script can be oblivious to the concept existing. That's however substantial effort and design for which we don't have a good second use case right now, so we'll likely get it wrong if we try.

If the `SchemaRegistry` of NativeScript would always return `null`, then they don't have to do anything else. Their renderer would also not change (see above).

> In any case, I'm unconvinced that that would be even worth it. Yes, it's a bit ugly that the word SecurityContext appears for Native Script, and that something in the template compiler talks about the DOM. But on the other hand, this just means Native Script has to return SecurityContext.NONE at one point, and bind a no-op implementation for SanitizationService. That seems like a perfectly fine tradeoff in exchange for the ability to do static security analysis. The layering violation really is not that bad (IMHO much less than the CompileIdentifierMetadata thing!).

This would be the first case where we purposefully ignore the platform separation of concepts in Angular, that we tried to keep for everything else so far. I don't see the need for doing this now, as the proposed changes above are comparatively small:

The proposal above is only adding interfaces above the implementation in this PR, and instead of generating code that specifically calls `sanitizeHtml`, `sanitizeUrl`, ... it would call `sanitize(DomSecurityContext.HTM, ...)`, `sanitize(DomSecurityContext.URL, ...)`.

**Comment by tbosch** [comment]
@mhevery We need you to vote in which direction this PR should go.

**Comment by mprobst** [comment]
Thinking a bit more about this, our goal is to not have DOM / browser specifics stand out in the Angular API on the template compiler / code generator level. Security as a concern is browser specific, so that means ideally we want no mention of a `SecurityContext` & Co on that level at all.

However all proposals we have, short of implementing a target specific code generator (or some similar abstraction in it), cause such problems - there is always some code in the package that we only have because of security, so we can move it around as much as we want, there's no clean solution.

Now if we keep the security an internal implementation detail of the package, we can just ignore it, or refactor later as we please - we won't break users. And we don't run into all the problems mentioned above.

So that makes the goal to limit the API leakage. To confirm - is that what we're after?

With the approach above, the only bit that leaks is knowledge of `SecurityContext` itself into `ElementSchemaRegistry`. AFAICT everything else is self contained. `ViewUtils` is private, the template code is a private API, the generated code is effectively private, and so on.

What about this approach:
- Only create the `securityContext` method in `DomElementSchemaRegistry`
- Check for `instanceof DomElementSchemaRegistry` in the code generator, and if it's there, get a security context (but the DOM specific enum), pipe through as before
- Only export `SecurityContext` and `SanitizationService` from the browser-specific packages
- `SanitizationService` is bound together with `ViewUtils` in `APPLICATION_COMMON_PROVIDERS` in core, but otherwise a private API as far as the compiler package and core are concerned.

**Comment by tbosch** [comment]
Regarding what to put where:
- platform independent interfaces that are referrred to from the generated code: into `core/...`, maybe `core/security`
- platform dependent code that is referred to from generated code: into `platform/{dom, ...}`

**Comment by mprobst** [comment]
Discussed offline, to keep everyone on the same page here's what we came up with:
- Less indirection would be fine for simplicity, but indirection seems manageable, and API complexity for schema registry is acceptable
- If introducing an indirection (`CompileIdentifierMetadata`), we might as well go all the way and generate direct calls to a static `safeHtml(value)` function
- Indirecting through a `SanitizationService` allows to configure sanitizer, but unclear why that'd be needed. If users want different sanitizers/safe html objects, they could configure the codegen/template compiler to use different schema element registries that refer to different sanitizers (though there's no easy mechanism to configure that at the moment, and it's pushing complexity to high end users).
- Make the sanitizer interface as seen from the parser and `ViewUtils` minimal, e.g. just a single `sanitize(context, value)` method. Also reduces coupling in codegen.
- Expose complete interface from browser package.
- Error reporting might get a bit ugly in the template parser, but manageable (e.g. multiple returns, throw with immediate catch)

Identifying a concrete enum value looks like this:

```
 new CompileIdentifierMetadata({
  name: 'DomSecurityContext.HTML',
  moduleUrl: 'angular2/src/platform/dom/DomSecurityContext',
  runtime: DomSecyrityContext.HTML
  });
```

... and similarly for a method.

Re whether an indirection/configurability is necessary for `SanitizationService` or not, I'll take a look tomorrow. Current hunch: we had configuration possibilities in AngularJS 1.x for things like what resource URLs are whitelisted, what protocols are legal for hyperlinks, and I could imagine whitelisting what CSP policies are in place (or generating CSP nonces for new `<style>` tags, etc). So I'd lean to keep the sanitizer indirection.

**Comment by mprobst** [comment]
Update: the `CompileIdentifierMetadata` indirection doesn't work, `TemplateParser` has to set security context to `SecurityContext.STYLE` for `[style.*]` bindings. I'm going to go with the `SecurityContext`. I won't pursue this any further for the time being, but will implement the API reduction (`sanitize(ctx, value)`).

**Pull request closed without merging**